observability/test/run_o11y_tests.py

```diff
@@ -0,0 +1,1174 @@
+        metric_client = monitoring_v3.MetricServiceClient()
+        interval = monitoring_v3.TimeInterval({
+            'end_time': {'seconds': int(time.time()),
+                         'nanos': test_impl.test_run_metadata.nanos},
+            'start_time': {'seconds': test_impl.test_run_metadata.script_start_seconds,
+                           'nanos': test_impl.test_run_metadata.nanos},
+        })
+        metrics_results: Dict[str, List[ListTimeSeriesResponse]] = {}
+        for metric_name in SUPPORTED_METRICS:
+            method_label = 'grpc_client_method' if 'client' in metric_name else 'grpc_server_method'
+            # cloud monitoring API only allows querying one metric at a time
+            filter_str = ('metric.type = "%s" AND metric.labels.identifier = "%s" ' \
+                          'AND metric.labels.%s = starts_with("grpc.testing.TestService")'
+                          % (metric_name, test_impl.identifier, method_label))
+            logger.info('Querying list_time_series: %s' % filter_str)
+            time_series = metric_client.list_time_series(
+                name=f'projects/{PROJECT}',
+                filter=filter_str,
+                interval=interval,
+            )
+            metrics_results[metric_name] = []
+            for series in time_series:
+                metrics_results[metric_name].append(series)
+            logger.debug('Found %d time_series for %s' % (
+                len(metrics_results[metric_name]), metric_name))
+        return cls(metrics_results)
+
+    def __init__(self, results: Dict[str, List[ListTimeSeriesResponse]]) -> None:
+        super().__init__()
+        self.results = results
+
+    def copy_to_dict(self, obj: Dict) -> Dict:
+        return {key: value for key, value in obj.items()}
+
+    def count_total(self, metric_name: str) -> int:
+        return len(self.results[metric_name])
+
+    def test_time_series_zero(self, metric_name: str) -> None:
+        num_metrics_result = self.count_total(metric_name)
+        self.assertEqual(num_metrics_result, 0)
+
+    def test_time_series_at_least_one(self, metric_name: str) -> None:
+        num_metrics_result = self.count_total(metric_name)
+        self.assertGreater(num_metrics_result, 0)
+
+    def test_metric_resource_type(self, metric_name: str, resource_type: str) -> None:
+        if self.count_total(metric_name) == 0:
+            return
+        for result in self.results[metric_name]:
+            self.assertEqual(result.resource.type, resource_type)
+
+    def test_metric_resource_labels(self, metric_name: str, resource_labels: Dict[str, str]) -> None:
+        if self.count_total(metric_name) == 0:
+            return
+        for result in self.results[metric_name]:
+            logger.debug('Metric resource labels: %s' % str(result.resource.labels.items()))
+            actual_labels = self.copy_to_dict(result.resource.labels)
```

Comment:
I think I did this because some of the fields returned by these cloud API calls are protobuf objects, which makes some of those deep comparisons I was trying to do a bit funky. I added a TODO to try to fix this in the next round.